core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala

```diff
@@ -0,0 +1,843 @@
+import org.apache.kafka.common.utils.Utils
+
+import kafka.utils._
+import kafka.common._
+import kafka.message._
+import kafka.log.FileMessageSet
+import kafka.metrics.KafkaMetricsGroup
+import kafka.common.TopicAndPartition
+import kafka.tools.MessageFormatter
+import kafka.api.ProducerResponseStatus
+import kafka.server.ReplicaManager
+
+import scala.collection._
+import java.io.PrintStream
+import java.nio.ByteBuffer
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.TimeUnit
+
+import com.yammer.metrics.core.Gauge
+
+class GroupMetadataManager(val config: OffsetConfig,
+                           replicaManager: ReplicaManager,
+                           zkUtils: ZkUtils,
+                           scheduler: Scheduler) extends Logging with KafkaMetricsGroup {
+
+  /* offsets cache */
+  private val offsetsCache = new Pool[GroupTopicPartition, OffsetAndMetadata]
+
+  /* group metadata cache */
+  private val groupsCache = new Pool[String, GroupMetadata]
+
+  /* partitions of consumer groups that are being loaded */
+  private val loadingPartitions: mutable.Set[Int] = mutable.Set()
+
+  /* lock for expiring stale offsets */
+  private val offsetExpireLock = new ReentrantReadWriteLock()
+
+  /* lock for removing offsets of a range partition */
+  private val offsetRemoveLock = new ReentrantReadWriteLock()
+
+  /* shutting down flag */
+  private val shuttingDown = new AtomicBoolean(false)
+
+  /* number of partitions for the consumer metadata topic */
+  private val groupMetadataTopicPartitionCount = getOffsetsTopicPartitionCount
+
+  this.logIdent = "[Offset Manager on Broker " + replicaManager.config.brokerId + "]: "
+
+  scheduler.schedule(name = "delete-expired-consumer-offsets",
+    fun = deleteExpiredOffsets,
+    period = config.offsetsRetentionCheckIntervalMs,
+    unit = TimeUnit.MILLISECONDS)
+
+  newGauge("NumOffsets",
+    new Gauge[Int] {
+      def value = offsetsCache.size
+    }
+  )
+
+  newGauge("NumGroups",
+    new Gauge[Int] {
+      def value = offsetsCache.keys.map(_.group).toSet.size
+    }
+  )
+
+  def partitionFor(group: String): Int = Utils.abs(group.hashCode) % groupMetadataTopicPartitionCount
+
+  def partitionLeaderIsLocal(partition: Int): Boolean = getHighWatermark(partition) != -1L
+
+  def partitionLoadingInProgress(partition: Int): Boolean = loadingPartitions synchronized loadingPartitions.contains(partition)
+
+  /**
+   * Get the group associated with the given groupId, or null if not found
+   */
+  def getGroup(groupId: String) = {
+      groupsCache.get(groupId)
+  }
+
+  /**
+   * Add a group or get the group associated with the given groupId if it already exists
+   */
+  def addGroup(groupId: String, protocolType: String): GroupMetadata = {
+    addGroup(groupId, new GroupMetadata(groupId, protocolType))
+  }
+
+  private def addGroup(groupId: String, group: GroupMetadata): GroupMetadata = {
+    groupsCache.put(groupId, group)
```

Comment:
Right. Basically I think there are two things we need to watch out for:
1. making sure the GroupMetadata is consistent. This happens with the GroupCoordinator using the GroupMetadata's implicit lock on any sort of state access or modifications to the group.
2. making sure the mapping holding the GroupMetadata is consistent. This happens through careful usage of the Pool (like using Pool.putIfNotExists in addGroup).